Implement sorensen_dice algorithm

theotff opened the pull request.

Commit:
Merge branch 'main' into api

Commit:
Implement sorensen algorithm

theotff commented:
Сделал привязку алгоритма соренсена. Поправил алгоритм так, как сказал @LiceyMaxim

ARQtty commented on api/utils/scoring/sorensen.py:
```diff
@@ -0,0 +1,10 @@
+    async def algorithm(self, code: str, lab: str) -> float:
+        return sorensens_dice.comparison(code, lab)
+
+    def name(self) -> str:
```
это можно вынести в базовый класс

theotff replied:
метод алгоритма?

ARQtty commented on models/sorensen/sorensens_dice.py:
```diff
@@ -0,0 +1,9 @@
+def comparison(first_phrase: str, second_phrase: str) -> float:
```
Раньше у нас был план сделать базовый класс алгоритма проверки. Сейчас добавился третий алгоритм сравнения, и всё начало быть несистемно и хаотично. Сделай пожалуйста базовый класс проверки в `models/` и отнаследуй от него все текущие алгоритмы сравнения

ARQtty reviewed (changes requested):
Нужен рефактор алгоритмов сравнения

Commit:
Refactor algorithms

- Create BaseAlgorithm
- Inherit algorithms from it
- Rename test files to single standart

theotff commented:
- Вынес __algorithm__ в базовый класс
- Сделал рефактор алгоритмов сравнения
  - Сделал базовый класс алгоритма
  - Отнаследовал все текущие алгоритмы сравнения от него
- Переименовал тесты по единому образцу

ARQtty commented on api/utils/scoring/base_check.py:
```diff
@@ -4,14 +4,16 @@
 class BaseCheck():
 
-    def __init__(self) -> None:
+    def __init__(self, comparison_method: BaseAlgorithm.comparison) -> None:
```
не стоит прокидывать конкретный метод класса. Достаточно прокинуть класс

ARQtty commented on api/utils/scoring/base_check.py:
```diff
@@ -42,7 +43,7 @@ async def check(self, lab_id: int) -> list:
-        raise NotImplementedError
+        return self.comparison_method(code, lab)
 
     def name(self) -> str:
```
тут можно сделать магию вроде `return self.__class__.__name__`

ARQtty commented on api/utils/scoring/base_check.py:
```diff
@@ -4,14 +4,16 @@
+from models.base_algorithm import BaseAlgorithm
 
 
 class BaseCheck():
```
Базовая Проверка? Что за проверка? Вне контекста непонятно. Базовый Алгоритм Сравнения был бы более трактуем

ARQtty commented on api/utils/scoring/jaccard.py:
```diff
@@ -1,10 +1,12 @@
 
     def name(self) -> str:
-        return self.__class__.__name__.lower()
+        return self.__class__.__name__.lower()[:-5]
```
хочется не заниматься таким переопределением. Можно делать это в базовом классе

ARQtty commented on models/base_algorithm.py:
```diff
@@ -0,0 +1,4 @@
+class BaseAlgorithm():
```
Очень хитро то, что здесь есть Базовый Алгоритм, и в бэке есть отдельный базовый алгоритм. Понятнее будет иметь только один (лучше сохранить алгоритмы в models/)

В бэке нужно описать бизнес-логику преобразования файлов с кодом в нужные алгоритму сравнения строки. Такое преобразование можно сделать отдельным классом. Тогда в бэке не будет алгоритмов сравнений, и архитектура станет чище

ARQtty reviewed (changes requested):
Базовый Алгоритм должен быть только один. Нужно переместить бизнес-логику работы с _файлами_ лабы в специальный для этого класс, а сравнение лаб как _строк_ оставить в models/

Commit:
Remove algorithms from api

- Move algorithm implementation to models

theotff commented:
Перекинул обработку алгоритмов в models

ARQtty commented on api/utils/lab_processing.py:
```diff
@@ -39,4 +40,30 @@ async def get_lab_code(self, path: Path, ext: str) -> str:
\ No newline at end of file
+        return code
+
+    async def get_labs(self, lab_id: int) -> List[tuple]:
```
tuple из чего?

ARQtty commented on api/utils/lab_processing.py:
```diff
@@ -39,4 +40,30 @@ async def get_lab_code(self, path: Path, ext: str) -> str:
+            lab_folder, lab_files = [file for file in os.walk(Path.joinpath(self.base_path, 'labs'))][-1][::2]
+            lab_expr = re.compile(r'.*\.(zip){1}$')
+
+            for lab_file in lab_files:
```
внутренность цикла хотелось бы вынести в отдельную функцию

ARQtty commented on api/utils/lab_scoring_handler.py:
```diff
@@ -6,19 +6,19 @@
         self.connector = SQLiteConnector()
-        self.levenshtein = Levenshtein()
-        self.jaccard = Jaccard()
         self.methods = [
```
тут не хватает коммента о том, зачем сохранять именно такие туплы

ARQtty commented on models/jaccard.py:
```diff
@@ -0,0 +1,16 @@
+
+        return len(shared) / len(str1_set.union(str2_set))
+
+    def name(self) -> str:
```
метод можно вынести в BaseAlgorithm, чтобы не определять в каждом наследнике

Commit:
Type hints, comments & cleanup

theotff commented:
- Убрал name в BaseAlgorithm
- Добавил hint для tuple в get_labs
- Добавил коммент про булеан
- Убрал обработку лабы из цикла в отдельную функцию

ARQtty commented on api/utils/lab_processing.py:
```diff
@@ -39,4 +40,36 @@ async def get_lab_code(self, path: Path, ext: str) -> str:
\ No newline at end of file
+        return code
+
+    async def get_comparison_lab(self, lab_file: str, lab_folder: str, lab_info: dict) -> Union[dict, None]:
```
вместо dict стоит использовать Dict[type1, type2]

ARQtty commented on api/utils/lab_processing.py:
```diff
@@ -39,4 +40,36 @@ async def get_lab_code(self, path: Path, ext: str) -> str:
+
+        return None
+
+    async def get_labs(self, lab_id: int) -> List[Tuple[str, dict]]:
```
Dict

ARQtty commented on models/base_algorithm.py:
```diff
@@ -0,0 +1,23 @@
+    def __init__(self) -> None:
+        self.lab_processing = LabProcessing()
+
+    async def check(self, lab_id: int):
```
возвращаемый тип

ARQtty reviewed (changes requested):
гуд. Чуть уточнить типы - и можно мержить

Commit:
Fix type hints

theotff commented:
Поправил типизацию

Commit:
Use joinpath instead of f-string
